packages/bsky/src/notifications.ts

```diff
@@ -0,0 +1,285 @@
+          notifications: notifications,
+        },
+        {
+          headers: {
+            'Content-Type': 'application/json',
+            accept: 'application/json',
+          },
+        },
+      )
+    } catch (error) {
+      throw new Error('Failed to send push notification')
+    }
+  }
+
+  async registerDeviceForPushNotifications(
+    did: string,
+    platform: Platform,
+    token: string,
+    endpoint = PUSH_NOTIF_SERVER_URL,
+    appId = BSKY_APP_ID,
+  ) {
+    try {
+      // check if token did pair already exists
+      const existing = await this.db.db
+        .selectFrom('notification_push_token')
+        .where('did', '=', did)
+        .where('token', '=', token)
+        .selectAll()
+        .executeTakeFirst()
+      if (existing) {
+        return
+      }
+
+      // if token doesn't exist, insert it
+      await this.db.db
+        .insertInto('notification_push_token')
+        .values({
+          did,
+          token,
+          platform,
+          endpoint,
+          appId,
+        })
+        .execute()
+    } catch (error) {
+      throw new Error('Failed to insert notification token')
+    }
+  }
+
+  async registerPushNotificationsEndpoint() {}
+
+  async registerPushNotificationsToken() {}
+
+  async getNotificationDisplayAttributes(notif: InsertableNotif) {
+    const {
+      author: authorDid,
+      reason,
+      reasonSubject: subjectUri, // if like/reply/quote/emtion, the post which was liked/replied to/mention is in/or quoted. if custom feed liked, the feed which was liked
+      recordUri,
+    } = notif
+    // 1. Get author's display name
+    const displayName = await this.db.db
+      .selectFrom('profile')
+      .where('creator', '=', authorDid)
+      .select(['displayName'])
+      .executeTakeFirst()
+    if (!displayName || !displayName?.displayName) {
+      throw new Error('Failed to get display name. User has no profile')
+    }
+    const author = NotificationServer.sanitizeDisplayName(
+      displayName.displayName,
+    )
+
+    // 2. Get post data content
+    // if reply, quote, or mention, get URI of the postRecord
+    // if like, or custom feed like, or repost get the URI of the reasonSubject
+    // if follow, get the URI of the author's profile
+    let title = ''
+    let body = ''
+
+    // check follow first and mention first because they don't have subjectUri and return
+    if (reason === 'follow') {
+      title = 'New follower!'
+      body = `${author} has followed you`
+      return { title, body }
+    } else if (reason === 'mention') {
+      title = `${author} mentioned you`
+      const mentionedPostData = await this.db.db
+        .selectFrom('post')
+        .where('uri', '=', recordUri)
+        .selectAll()
+        .executeTakeFirst()
+      body = mentionedPostData?.text || ''
+      return { title, body }
+    }
+
+    if (!subjectUri) {
+      throw new Error('Failed to get subject URI')
+    }
+
+    const postData = await this.db.db
```

Comment:
since the indexing service already has access to the record in question, it may be better to just get the relevant data from there instead of following it up with a bunch of selects

especially a Create with a wide fanout - like a reply in a thread - could end up hitting the database a lot

Reply:
ah just realized - i guess it doesn't have access to the subject which would be the big saver 🤔

Reply:
still let's try to do as few selects as possible for the notif fanout

Reply:
yeah... it would...

Reply:
I'd really like to eliminate the DB hits here 🤔 

can we just use `handle` instead of `displayName` and thread the record through to this method so we don't have to hit the db?